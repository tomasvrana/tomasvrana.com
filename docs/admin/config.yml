backend:
  name: git-gateway
  branch: staging
  accept_roles:
    - admin
    - editor
media_folder: static/images/uploads
public_folder: /images/uploads
publish_mode: editorial_workflow
show_preview_links: true
collections:
  - name: meta_data
    label: Meta data
    folder: resources/content/meta
    create: false
    fields:
      - label: "Title"
        name: "title"
        widget: "string"
      - name: head
        label: SEO <head /> data
        widget: object
        fields:
          - name: title
            label: Site title tag value
            widget: string
          - name: meta
            label: Meta tags
            label_singular: Meta tag
            widget: list
            fields:
              - name: name
                label: name attribute
                widget: string
              - name: content
                label: Content attribute, if no value is given then value will be taken from "File"
                widget: string
                required: false
                default: ""
              - name: file
                label: File
                description: Used instead of content where the meta tag is a URL to a file such as image etc.
                required: false
                default: ""
                widget: file
                allow_multiple: false
  - name: header
    label: Header
    folder: resources/content/header
    fields:
      - label: "Title"
        name: "title"
        widget: "string"
      - name: navigation
        label: Navigation
        widget: object
        fields:
          - name: labels
            label: Navigation links text
            widget: object
            fields:
              - name: home
                label: Home page
                widget: string
              - name: page_2
                label: Page 2
                widget: string
  - name: footer
    label: Footer
    folder: resources/content/footer
    create: false
    fields:
      - label: "Title"
        name: "title"
        widget: "string"
      - label: cardano_info
        name: Cardano info heading
        widget: string
      - label: Community email address
        name: community_email
        widget: string
      - name: community_info
        label: Community info heading
        widget: string
      - label: Cardano informational links
        label_singular: Cardano informational link
        name: cardano_links
        widget: list
        fields:
          - label: URL
            name: href
            widget: string
          - label: Label
            name: label
            widget: string
      - label: Cardano community links
        label_singular: Cardano community link
        name: community_links
        widget: list
        fields:
          - label: URL
            name: href
            widget: string
          - label: Label
            name: label
            widget: string
      - label: Footer text
        name: body
        description: Text blob in footer, parsed as HTML
        widget: string
  - name: home_page
    label: Home page
    folder: resources/content/pages/index
    create: false
    fields:
      - label: "Title"
        name: "title"
        widget: "string"
      - name: content
        label: Home page content
        widget: object
        fields:
          - name: title
            label: Main title
            widget: string
          - name: page_2_link
            label: Page 2 link text
            widget: string
  - name: 404_page
    label: 404 page
    folder: resources/content/pages/404
    create: false
    fields:
      - label: "Title"
        name: "title"
        widget: "string"
      - name: content
        label: 404 page content
        widget: object
        fields:
          - name: title
            label: Title
            widget: string
          - name: subtitle
            label: Subtitle
            widget: string
          - name: description
            label: description
            widget: string
