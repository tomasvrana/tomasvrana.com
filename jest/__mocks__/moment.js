const moment = jest.requireActual('moment')

module.exports = {
  ...moment,
  locale: jest.fn()
}
