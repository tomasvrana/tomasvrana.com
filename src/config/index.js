import siteConfig from '../../site.config.json'

export default {
  ...siteConfig,
  ga: { trackingID: 'G-8Q6SN0CZ6Y' }
}
