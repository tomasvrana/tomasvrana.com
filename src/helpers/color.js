import TinyColor from '@ctrl/tinycolor'
import logger from './logger'

const makeTransparent = (color, alpha) => {
  try {
    const c = new TinyColor(color)
    c.setAlpha(alpha)
    return c.toRgbString()
  } catch (error) {
    logger.error({
      description: 'error making color transparent',
      args: [ color, alpha ],
      error
    })
    return '#000000'
  }
}

export {
  makeTransparent
}
