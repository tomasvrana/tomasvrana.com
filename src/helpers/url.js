import { triggerEvent } from './dom'
import config from '../config'

const window = global.window || { location: { pathname: '/', href: '', origin: '', hash: '', hostname: '' }, history: {} }

export const getURIPath = () => window.location.pathname
export const getURIPathWithoutLang = lang => getURIPath().replace(new RegExp(`^\\/${lang}`), '')
export const languageSetInURL = (url = null) => !!config.availableLanguages[(url || getURIPath()).split('/').splice(1)[0]]
export const getHref = (path = null) => path ? `${window.location.origin}${path}` : window.location.href
export const setHref = href => (window.location.href = href)
export const isRelative = url => !url.match(/^(\/\/|https?:\/\/|mailto:)/)
export const isPath = url => !url.match(/^(\/|\/[^\\/].*)$/)
export const getHash = () => window.location.hash
export const getHostname = () => window.location.hostname

const getHashParams = () => {
  const currentHash = getHash().replace(/^#/, '')
  const params = currentHash.split('&').map(p => ({ [p.split('=')[0]]: p.split('=')[1] }))
  let paramsObj = {}
  params.forEach(p => (paramsObj = { ...paramsObj, ...p }))
  return paramsObj
}

export const setHashParam = (key, value) => {
  const params = getHashParams()
  if (!value) {
    delete params[key]
  } else {
    params[key] = value
  }

  const paramStrings = Object.keys(params).map(k => params[k] && k ? `${k}=${params[k]}` : '').filter(v => !!v)
  if (paramStrings.length > 0) {
    if (window.history.pushState) {
      window.history.pushState(null, null, `#${paramStrings.join('=')}`)
      triggerEvent({ type: 'hashchange' })
    } else {
      window.location.hash = `#${paramStrings.join('=')}`
    }
  } else {
    if (window.history.pushState) {
      window.history.pushState(null, null, '#')
      triggerEvent({ type: 'hashchange' })
    } else {
      window.location.hash = ''
    }
  }
}

export const getHashParam = key => {
  return getHashParams()[key] || ''
}
