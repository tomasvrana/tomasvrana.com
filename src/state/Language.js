import React, { createContext, useState, useEffect } from 'react'
import { navigate } from 'gatsby'
import PropTypes from 'prop-types'
import config from '../config'
import * as moment from 'moment'
import { getURIPath, getHash, languageSetInURL } from '../helpers/url'
import { getLocalStorageValue, setLocalStorageValue } from '../helpers/localStorage'
import { getNavigatorLanguage } from '../helpers/navigator'
import { autoCapture } from '../helpers/analytics'
import { LANGUAGE } from '../constants/analytics'

const LanguageContext = createContext()
const Consumer = LanguageContext.Consumer

function getUrlLang () {
  const [ urlLang ] = getURIPath().split('/').splice(1)
  return urlLang
}

function getDefaultLanguage () {
  // Use URL language where available
  if (config.availableLanguages[getUrlLang()]) return getUrlLang()
  // Fallback to local storage or default config
  const localStorageLang = getLocalStorageValue('lang')
  // Users browser language
  const navigatorLanguage = getNavigatorLanguage().substring(0, 2)
  // Use local storage first as they have already visited the site
  if (localStorageLang && config.availableLanguages[localStorageLang]) return localStorageLang
  // Use the users browser language if it supported by the site
  if (navigatorLanguage && config.availableLanguages[navigatorLanguage]) return navigatorLanguage
  // Fall back to the first available language in the config
  return Object.keys(config.availableLanguages)[0]
}

const Provider = ({ children }) => {
  const [ lang, updateLang ] = useState(getDefaultLanguage())
  const setLang = language => {
    if (!config.availableLanguages[language]) throw new Error(`Invalid language ${language}`)
    const pathParts = getURIPath().split('/').splice(1)
    if (languageSetInURL()) {
      pathParts.splice(0, 1, language)
    } else {
      pathParts.unshift(language)
    }

    setLocalStorageValue('lang', language)
    moment.locale(language)
    updateLang(language)
    if (lang !== language) autoCapture({ category: LANGUAGE, label: language, action: 'updated_language' })
    navigate(`/${pathParts.join('/')}${getHash()}`)
  }

  useEffect(() => {
    setLang(lang)
  }, [])

  return (
    <LanguageContext.Provider
      value={{
        lang,
        setLang,
        availableLocales: Object.keys(config.availableLanguages).map(key => config.availableLanguages[key].locale),
        locale: config.availableLanguages[lang].locale
      }}
    >
      {children}
    </LanguageContext.Provider>
  )
}

Provider.propTypes = {
  children: PropTypes.any
}

export {
  Provider,
  Consumer
}
