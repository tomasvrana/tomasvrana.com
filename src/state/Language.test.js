import React from 'react'
import * as gatsby from 'gatsby'
import { mount } from 'enzyme'
import { Provider, Consumer } from './Language'
import * as moment from 'moment'
import * as url from '../helpers/url'
import * as localStorage from '../helpers/localStorage'
import * as navigator from '../helpers/navigator'
import * as analytics from '../helpers/analytics'
import config from '../config'
import { LANGUAGE } from '../constants/analytics'

const getProvider = props => {
  return mount(
    <Provider
      children={(
        <Consumer>
          {({ lang, locale, availableLocales }) => (
            <div>
              <p key='current-lang'>{lang}</p>
              <p key='current-locale'>{locale}</p>
              <ul key='available-locales'>
                {availableLocales.map(locale => (
                  <li key={locale}>{locale}</li>
                ))}
              </ul>
            </div>
          )}
        </Consumer>
      )}
      {...props}
    />
  )
}

const availableLanguages = {
  en: { label: 'English', locale: 'en_GB' },
  ja: { label: 'Japanese', locale: 'ja_JA' },
  fr: { label: 'French', locale: 'fr_FR' }
}

describe('Language', () => {
  let originalAvailableLanguages
  beforeEach(() => {
    originalAvailableLanguages = { ...config.availableLanguages }
    url.getURIPath = jest.fn(() => '/')
    url.getHash = jest.fn(() => '')
    url.languageSetInURL = jest.fn()
    localStorage.getLocalStorageValue = jest.fn(() => '')
    localStorage.setLocalStorageValue = jest.fn()
    navigator.getNavigatorLanguage = jest.fn(() => '')
    analytics.autoCapture = jest.fn()
    config.availableLanguages = availableLanguages
  })

  afterEach(() => {
    config.availableLanguages = originalAvailableLanguages
  })

  test('the Provider renders correctly', () => {
    expect(getProvider()).toMatchSnapshot()
  })

  describe('when the component "mounts"', () => {
    let expectedLanguage
    beforeEach(() => {
      getProvider()
      expectedLanguage = Object.keys(availableLanguages).shift()
    })

    test('it sets the local storage language', () => {
      expect(localStorage.setLocalStorageValue).toBeCalledWith('lang', expectedLanguage)
    })

    test('it sets the moment locale', () => {
      expect(moment.locale).toBeCalledWith(expectedLanguage)
    })

    test('it does not track a language change', () => {
      expect(analytics.autoCapture).not.toBeCalled()
    })

    test('it navigates to the new URL', () => {
      expect(gatsby.navigate).toBeCalledWith(`/${expectedLanguage}/`)
    })
  })

  describe('when the language is in the URL', () => {
    describe('when the language is supported', () => {
      beforeEach(() => {
        url.getURIPath.mockReturnValue(`/${Object.keys(availableLanguages).pop()}/`)
      })

      test(`it renders with the '${Object.keys(availableLanguages).pop()}' language`, () => {
        expect(getProvider()).toMatchSnapshot()
      })
    })

    describe('when the language is not supported', () => {
      beforeEach(() => {
        url.getURIPath.mockReturnValue('/aaa/')
      })

      test(`it renders with the first supported language ('${Object.keys(availableLanguages)[0]}')`, () => {
        expect(getProvider()).toMatchSnapshot()
      })
    })
  })

  describe('when the language is set in local storage', () => {
    describe('when the language is supported', () => {
      beforeEach(() => {
        localStorage.getLocalStorageValue.mockImplementation(key => key === 'lang' && Object.keys(availableLanguages).pop())
      })

      test(`it renders with the '${Object.keys(availableLanguages).pop()}' language`, () => {
        expect(getProvider()).toMatchSnapshot()
      })
    })

    describe('when the language is not supported', () => {
      beforeEach(() => {
        localStorage.getLocalStorageValue.mockImplementation(key => key === 'lang' && 'aaa')
      })

      test(`it renders with the first supported language ('${Object.keys(availableLanguages)[0]}')`, () => {
        expect(getProvider()).toMatchSnapshot()
      })
    })
  })

  describe('when there is a language in the navigator', () => {
    describe('when the language is supported', () => {
      beforeEach(() => {
        navigator.getNavigatorLanguage.mockReturnValue(Object.keys(availableLanguages).pop())
      })

      test(`it renders with the '${Object.keys(availableLanguages).pop()}' language`, () => {
        expect(getProvider()).toMatchSnapshot()
      })
    })

    describe('when the language is not supported', () => {
      beforeEach(() => {
        navigator.getNavigatorLanguage.mockReturnValue('aaa')
      })

      test(`it renders with the first supported language ('${Object.keys(availableLanguages)[0]}')`, () => {
        expect(getProvider()).toMatchSnapshot()
      })
    })
  })

  describe('updating the language', () => {
    let wrapper, language
    beforeEach(() => {
      language = {}
      wrapper = getProvider({
        children: (
          <Consumer>
            {({ lang, locale, availableLocales, setLang }) => (
              <div>
                <p key='current-lang'>{lang}</p>
                <p key='current-locale'>{locale}</p>
                <ul key='available-locales'>
                  {availableLocales.map(locale => (
                    <li key={locale}>{locale}</li>
                  ))}
                </ul>
                <button onClick={() => setLang(language.language)}>Update language</button>
              </div>
            )}
          </Consumer>
        )
      })
    })

    describe('updating to an unsupported language', () => {
      beforeEach(() => {
        language.language = 'aaa'
      })

      test('an error is thrown', () => {
        expect(() => {
          wrapper.find('button').simulate('click')
        }).toThrowError(`Invalid language ${language.language}`)
      })
    })

    describe('updating to a supported language', () => {
      beforeEach(() => {
        language.language = Object.keys(availableLanguages)[1]
        wrapper.find('button').simulate('click')
      })

      test('it sets the local storage language', () => {
        expect(localStorage.setLocalStorageValue).toBeCalledWith('lang', language.language)
      })

      test('it sets the moment locale', () => {
        expect(moment.locale).toBeCalledWith(language.language)
      })

      test('it tracks the language change', () => {
        expect(analytics.autoCapture).toBeCalledWith({ category: LANGUAGE, label: language.language, action: 'updated_language' })
      })

      test('it navigates to the new URL', () => {
        expect(gatsby.navigate).toBeCalledWith(`/${language.language}/`)
      })
    })
  })
})
