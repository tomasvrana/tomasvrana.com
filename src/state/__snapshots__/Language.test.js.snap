// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Language the Provider renders correctly 1`] = `
<Provider>
  <div>
    <p
      key="current-lang"
    >
      en
    </p>
    <p
      key="current-locale"
    >
      en_GB
    </p>
    <ul
      key="available-locales"
    >
      <li
        key="en_GB"
      >
        en_GB
      </li>
      <li
        key="ja_JA"
      >
        ja_JA
      </li>
      <li
        key="fr_FR"
      >
        fr_FR
      </li>
    </ul>
  </div>
</Provider>
`;

exports[`Language when the language is in the URL when the language is not supported it renders with the first supported language ('en') 1`] = `
<Provider>
  <div>
    <p
      key="current-lang"
    >
      en
    </p>
    <p
      key="current-locale"
    >
      en_GB
    </p>
    <ul
      key="available-locales"
    >
      <li
        key="en_GB"
      >
        en_GB
      </li>
      <li
        key="ja_JA"
      >
        ja_JA
      </li>
      <li
        key="fr_FR"
      >
        fr_FR
      </li>
    </ul>
  </div>
</Provider>
`;

exports[`Language when the language is in the URL when the language is supported it renders with the 'fr' language 1`] = `
<Provider>
  <div>
    <p
      key="current-lang"
    >
      fr
    </p>
    <p
      key="current-locale"
    >
      fr_FR
    </p>
    <ul
      key="available-locales"
    >
      <li
        key="en_GB"
      >
        en_GB
      </li>
      <li
        key="ja_JA"
      >
        ja_JA
      </li>
      <li
        key="fr_FR"
      >
        fr_FR
      </li>
    </ul>
  </div>
</Provider>
`;

exports[`Language when the language is set in local storage when the language is not supported it renders with the first supported language ('en') 1`] = `
<Provider>
  <div>
    <p
      key="current-lang"
    >
      en
    </p>
    <p
      key="current-locale"
    >
      en_GB
    </p>
    <ul
      key="available-locales"
    >
      <li
        key="en_GB"
      >
        en_GB
      </li>
      <li
        key="ja_JA"
      >
        ja_JA
      </li>
      <li
        key="fr_FR"
      >
        fr_FR
      </li>
    </ul>
  </div>
</Provider>
`;

exports[`Language when the language is set in local storage when the language is supported it renders with the 'fr' language 1`] = `
<Provider>
  <div>
    <p
      key="current-lang"
    >
      fr
    </p>
    <p
      key="current-locale"
    >
      fr_FR
    </p>
    <ul
      key="available-locales"
    >
      <li
        key="en_GB"
      >
        en_GB
      </li>
      <li
        key="ja_JA"
      >
        ja_JA
      </li>
      <li
        key="fr_FR"
      >
        fr_FR
      </li>
    </ul>
  </div>
</Provider>
`;

exports[`Language when there is a language in the navigator when the language is not supported it renders with the first supported language ('en') 1`] = `
<Provider>
  <div>
    <p
      key="current-lang"
    >
      en
    </p>
    <p
      key="current-locale"
    >
      en_GB
    </p>
    <ul
      key="available-locales"
    >
      <li
        key="en_GB"
      >
        en_GB
      </li>
      <li
        key="ja_JA"
      >
        ja_JA
      </li>
      <li
        key="fr_FR"
      >
        fr_FR
      </li>
    </ul>
  </div>
</Provider>
`;

exports[`Language when there is a language in the navigator when the language is supported it renders with the 'fr' language 1`] = `
<Provider>
  <div>
    <p
      key="current-lang"
    >
      fr
    </p>
    <p
      key="current-locale"
    >
      fr_FR
    </p>
    <ul
      key="available-locales"
    >
      <li
        key="en_GB"
      >
        en_GB
      </li>
      <li
        key="ja_JA"
      >
        ja_JA
      </li>
      <li
        key="fr_FR"
      >
        fr_FR
      </li>
    </ul>
  </div>
</Provider>
`;
